src/NuGet.Core/NuGet.Versioning/VersionRange.cs

```diff
@@ -297,9 +298,12 @@ public VersionRange ToNonSnapshotRange()
             {
                 var minVersion = MinVersion;
 
-                if (Float.FloatBehavior == NuGetVersionFloatBehavior.Prerelease)
+                if (Float!.FloatBehavior == NuGetVersionFloatBehavior.Prerelease)
                 {
+#pragma warning disable CS8604 // Possible null reference argument.
+                    // Possible bug?
                     minVersion = GetNonSnapshotVersion(minVersion);
```

Comment:
Would MemberNotNullWhen help here as well?

Reply:
This was quite problematic. See this commit: https://github.com/NuGet/NuGet.Client/pull/5097/commits/92917270a9526c115f9914de784678c6aad7ae83

Firstly, `MemberNotNullWhenAttribute` appears to only allow you to reference members that are in the same class. `MinVersion` is defined in `VersionRangeBase`, whereas all float stuff is in `VersionRange`. Therefore, I had to add `public new NuGetVersion? MinVersion => base.MinVersion;`, but then all the annotations from `VersionRangeBase` that says `MinVersion` is not null when `IncludeLowerBound`, `HasMinVersion`, or `IncludeLowerAndUpperBound` is true. So, then I had to copy those properties, but since `IncludeLowerAndUpperBound` has `MemberNotNullWhen(true, nameof(MaxVersion))`, then I had to copy all the `MaxVersion` properties too.

Secondly, I added a check in the constructor so when `floatRange != null && floatRange.FloatBehaviour != null && minVersion == null` it throws, so that we ensure that the `MemberNotNullWhen(true, nameof(MinVersion))` is actually validated. However, the two obsolete properties `VersionRange.AllFloating` and `VersionRange.AllStableFloating` violated this assertion and threw exceptions. Changing the constructors to pass in `NuGetVersion.Parse("0.0.0")` as the min version caused the tests that assert the `ToNormalizedString` output changed.  Maybe I could add an internal contructor for the two obsolete static instances in `VersionRangeFactory` to avoid throwing, but then they violate the not-null annotation, and those code paths might throw exceptions.

If we care most about code-correctness, I think we should keep the commit. But it really feels bad.

The new public API changes in the commit could be removed if we delete the pointless `VersionRangeBase` abstraction, but unfortunately that's public, so would be a public API breaking change. That wouldn't solve the behaviour change in the obsolete `AllFloating` and `AllStableFloating` `ToNormalizedString` output changes.  If we're going to break public APIs, maybe we just delete those APIs, since they were marked obsolete in NuGet 5.5 in 2019. It would be nice to follow SemVer and wait until the next major NuGet version increment, but our last attempt at making breaking changes at a major version change wasn't successful.